src/cpp/fastdds/domain/DomainParticipantImpl.cpp

```diff
@@ -0,0 +1,722 @@
+#include <fastdds/domain/DomainParticipant.hpp>
+#include <fastdds/domain/DomainParticipantListener.hpp>
+#include <fastrtps/rtps/participant/ParticipantDiscoveryInfo.h>
+#include <fastrtps/rtps/reader/ReaderDiscoveryInfo.h>
+#include <fastrtps/rtps/writer/WriterDiscoveryInfo.h>
+#include <fastrtps/rtps/builtin/liveliness/WLP.h>
+
+#include <fastrtps/topic/TopicDataType.h>
+
+#include <fastrtps/rtps/participant/RTPSParticipant.h>
+
+#include <fastrtps/attributes/PublisherAttributes.h>
+#include "../publisher/PublisherImpl.hpp"
+#include <fastdds/publisher/Publisher.hpp>
+
+#include <fastrtps/attributes/SubscriberAttributes.h>
+#include "../subscriber/SubscriberImpl.hpp"
+#include <fastdds/subscriber/Subscriber.hpp>
+
+#include <fastdds/topic/DataReader.hpp>
+#include <fastdds/topic/DataWriter.hpp>
+
+#include <fastrtps/rtps/RTPSDomain.h>
+
+#include <fastrtps/transport/UDPv4Transport.h>
+#include <fastrtps/transport/UDPv6Transport.h>
+#include <fastrtps/transport/test_UDPv4Transport.h>
+
+#include <fastrtps/log/Log.h>
+
+#include <chrono>
+
+using namespace eprosima;
+using namespace eprosima::fastdds;
+
+using fastrtps::ParticipantAttributes;
+using fastrtps::TopicDataType;
+using fastrtps::rtps::RTPSDomain;
+using fastrtps::rtps::RTPSParticipant;
+using fastrtps::rtps::ParticipantDiscoveryInfo;
+#if HAVE_SECURITY
+using fastrtps::rtps::ParticipantAuthenticationInfo;
+#endif
+using fastrtps::rtps::ReaderDiscoveryInfo;
+using fastrtps::rtps::ReaderProxyData;
+using fastrtps::rtps::WriterDiscoveryInfo;
+using fastrtps::rtps::WriterProxyData;
+using fastrtps::rtps::InstanceHandle_t;
+using fastrtps::rtps::GUID_t;
+using fastrtps::rtps::EndpointKind_t;
+using fastrtps::rtps::ResourceEvent;
+using fastrtps::Log;
+
+DomainParticipantImpl::DomainParticipantImpl(
+        const ParticipantAttributes& patt,
+        DomainParticipant* pspart,
+        DomainParticipantListener* listen)
+    : att_(patt)
+    , rtps_participant_(nullptr)
+    , participant_(pspart)
+    , listener_(listen)
+#pragma warning (disable : 4355 )
+    , rtps_listener_(this)
+{
+    participant_->impl_ = this;
+}
+
+void DomainParticipantImpl::disable()
+{
+    participant_->set_listener(nullptr);
+    {
+        std::lock_guard<std::mutex> lock(mtx_pubs_);
+        for (auto pub_it = publishers_.begin(); pub_it != publishers_.end(); ++pub_it)
+        {
+            pub_it->second->disable();
+        }
+    }
+
+    {
+        std::lock_guard<std::mutex> lock(mtx_subs_);
+        for (auto sub_it = subscribers_.begin(); sub_it != subscribers_.end(); ++sub_it)
+        {
+            sub_it->second->disable();
+        }
+    }
+}
+
+DomainParticipantImpl::~DomainParticipantImpl()
+{
+    {
+        std::lock_guard<std::mutex> lock(mtx_pubs_);
+        for (auto pub_it = publishers_.begin(); pub_it != publishers_.end(); ++pub_it)
+        {
+            delete pub_it->second;
+        }
+        publishers_.clear();
+        publishers_by_handle_.clear();
```

Comment:
Since `lookup` methods use an `InstanceHandle_t` to look for a publisher or a subscriber, a second map working as an index for these methods can improve the performance with a minimal memory impact.

Reply:
OK